src/MagnumPlugins/ObjImporter/ObjImporter.cpp

```diff
@@ -28,81 +28,344 @@
+
+/* Not using PhongMaterialData, since we may want to set color and texture to
+ * later decide which flags we set. We do not know whether we have a texture
+ * or color beforehand. */
+struct ObjMaterial {
+    std::string name;
+    Color3 ambient;
+    Color3 diffuse;
+    Color3 specular;
+    Float specularity;
+    Int ambientTexture = -1;
+    Int diffuseTexture = -1;
+    Int specularTexture = -1;
+};
+
+struct ObjObject;
+struct ObjGroup;
+struct ObjMesh;
+
+struct ObjMeshData {
+    ObjGroup& group;
+    Int materialId; /* From 'usemtl' keyword */
+
+    ObjMesh* poInts = nullptr;
+    ObjMesh* lines = nullptr;
+    ObjMesh* faces = nullptr;
+
+    explicit ObjMeshData(ObjGroup& g, Int matId): group(g), materialId(matId) {}
+};
+
+struct ObjGroup {
+    ObjObject& object;
+    std::string name; /* From 'g' keyword */
+
+    std::vector<std::unique_ptr<ObjMeshData>> meshes;
+    std::unordered_map<Int, Int> meshPerMaterial;
+
+    explicit ObjGroup(ObjObject& o): object(o) {}
+
+    /* Create or get mesh for given material id */
+    ObjMeshData& meshDataForMaterial(Int materialId) {
+        if(meshPerMaterial.find(materialId) != meshPerMaterial.end()) {
+            return *meshes[meshPerMaterial[materialId]];
+        } else {
+            meshPerMaterial[materialId] = meshes.size();
+            meshes.emplace_back(new ObjMeshData{*this, materialId});
+            ObjMeshData& mesh = *meshes.back();
+            return mesh;
+        }
+    }
+};
+
+struct ObjObject {
+    std::string name; /* From 'o' keyword */
+    std::vector<std::unique_ptr<ObjGroup>> groups;
+
+    ObjObject(): name{""} {}
+    ObjObject(ArrayView<const char> name): name{name.data(), name.size()} {}
+};
+
+/* An Intermediate object representing a mesh and it's properties as well as
+ * where to find the data associated to it */
+struct ObjMesh {
+    ObjMeshData& data; /* Parent object containing data shared meshes for different primitives */
+    MeshPrimitive primitive;
+
+    /* Sections of the file belonging to this mesh */
+    std::vector<ArrayView<const char>> sections;
+
+    Int minPrimitives = 0; /* For smarter vector memory allocation later */
+
+    explicit ObjMesh(ObjMeshData& d, MeshPrimitive p): data(d), primitive(p) {}
+
+    std::string name() {
+        std::string name = data.group.object.name;
+
+        if(!data.group.name.empty()) {
+            name += ":" + data.group.name;
+        }
+
+        if(data.materialId != -1) {
+            name += "$" + std::to_string(data.materialId);
+        }
+
+        const Int numPrimitiveTypes =
+            ((data.poInts) ? 1 : 0)
+            + ((data.lines) ? 1 : 0)
+            + ((data.faces) ? 1 : 0);
+        if(numPrimitiveTypes > 1) {
+            std::ostringstream out;
+            Debug(&out) << primitive;
+            name += "%" + out.str();
+        }
+
+        return name;
+    }
+};
+
+
+/* The state of the imported generated by openData() */
+struct ImporterState {
```

Comment:
Took me a while to realize what's going on, sorry for the late reply.

> but those two added members in and fileRoot shouldn't be, because they are set before the state is reset

I think they don't need to be. One option:

```cpp
void ObjImporter::doOpenFile(const std::string& filename) {
    AbstractImporter::doOpenFile(filename);
    _state->fileRoot = Utility::Directory::path(filename);
}

void ObjImporter::doOpenData(Containers::ArrayView<const char> data) {
    _state.reset(new ImporterState);
    _state->in = Containers::Array<char>{data.size()};
    std::copy(data.begin(), data.end(), _state->in.begin());
    parse();
}
```

The above assumes that `parse()` doesn't need to access the `_fileRoot` parameter (which it shouldn't, if you defer the material parsing to later, as I suggested). Another option could be this, where the `_state` is only populated if it's not already there:

```cpp
void ObjImporter::doOpenFile(const std::string& filename) {
    _state.reset(new ImporterState);
    _state->fileRoot = Utility::Directory::path(filename);
    AbstractImporter::doOpenFile(filename);
}

void ObjImporter::doOpenData(Containers::ArrayView<const char> data) {
    if(!_state) _state.reset(new ImporterState);
    _state->in = Containers::Array<char>{data.size()};
    std::copy(data.begin(), data.end(), _state->in.begin());
    parse();
}
```

The `AbstractImporter::open*()` function implementation is [unconditionally calling](https://github.com/mosra/magnum/blob/2384d5cccdb020342ad9ad6d40d700c7586acbc1/src/Magnum/Trade/AbstractImporter.cpp#L56) `close()` before calling the `doOpen()` implementations, so you can be always sure that if `_state` is already populated, it's yours and not some stale one from a previous file.